utils/zerovec/src/ule/vartuple.rs

```diff
@@ -0,0 +1,439 @@
+    assert_eq!(
+        &var_tuple_ule.variable.variable,
+        ZeroSlice::from_ule_slice(&[b'I', b'C', b'U'])
+    );
+}
+
+#[test]
+fn test_vartuple_validation() {
+    let invalid0 = VarTuple {
+        sized: [0x80, 0x80, 0x80],
+        variable: "hello",
+    };
+    let invalid0_ule: Box<VarTupleULE<[u8; 3], str>> = super::encode_varule_to_box(&invalid0);
+    assert!(VarTupleULE::<char, str>::parse_byte_slice(invalid0_ule.as_byte_slice()).is_err());
+
+    let invalid1 = VarTuple {
+        sized: '🦙',
+        variable: b"\x80invalid".as_slice(),
+    };
+    let invalid1_ule: Box<VarTupleULE<char, [u8]>> = super::encode_varule_to_box(&invalid1);
+    assert!(VarTupleULE::<char, str>::parse_byte_slice(invalid1_ule.as_byte_slice()).is_err());
+}
+
+/// Errors associated with [`TinyVarVar`].
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum TinyVarVarError {
+    /// The value is too long to encode.
+    TooLong,
+}
+
+/// A sized type combining two types that can be converted to a [`TinyVarVarULE`].
+///
+/// The first of the two types is expected to be short (no longer than 127 bytes).
+/// There is no limit to the length of the second type.
+///
+/// # Examples
+///
+/// Store a default string and an array of other strings:
+///
+/// ```
+/// use zerovec::ule::vartuple::{TinyVarVar, TinyVarVarULE};
+/// use zerovec::{VarZeroVec, VarZeroSlice};
+///
+/// let homophones = [
+///     ("cense", &["cents", "scents", "sense"][..]),
+///     ("peak", &["peek", "peke", "pique"][..]),
+///     ("poor", &["pore", "pour"][..]),
+///     ("rain", &["reign", "rein"][..]),
+/// ];
+///
+/// let homophones_tvv = homophones
+///     .iter()
+///     .map(|(word, extras)| TinyVarVar::try_new(*word, *extras).unwrap())
+///     .collect::<Vec<_>>();
+///
+/// let homophones_vzv = VarZeroVec::<TinyVarVarULE<str, VarZeroSlice<str>>>::from(&homophones_tvv);
+///
+/// assert_eq!(homophones_vzv.len(), 4);
+/// assert_eq!(homophones_vzv.get(1).unwrap().get().0, "peak");
+/// assert_eq!(homophones_vzv.get(1).unwrap().get().1.get(0), Some("peek"));
+/// ```
+#[derive(Debug)]
+pub struct TinyVarVar<T0, T1, V0: VarULE + ?Sized, V1: VarULE + ?Sized>(
+    // Invariant: T0 encoded as V0 has a length that fits in 7 bits
+    T0,
+    T1,
+    PhantomData<V0>,
+    PhantomData<V1>,
+);
+
+/// A dynamically-sized VarULE type combining two dynamically-sized VarULE types.
+///
+/// The first of the two types is expected to be short (no longer than 127 bytes).
+/// There is no limit to the length of the second type.
+///
+/// See [`TinyVarVar`] for examples.
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct TinyVarVarULE<V0, V1>
+where
+    V0: VarULE + ?Sized,
+    V1: VarULE + ?Sized,
+{
+    _v0: PhantomData<V0>,
+    _v1: PhantomData<V1>,
+    // Invariants:
+    // - the slice begins with a byte indicating the cutoff between V0 and V1
+    // - the high bit of the first byte is 0 (reserved for a future varint)
```

Comment:
OK, I discovered after running this through full datagen that there is some Adlam text for currency long names or something that is long enough that it violates this invariant.

Should I:

1. Make everyone use `u16`
2. Implement a varint approach (I have one stashed)
3. Start using the high bit, which should cover the known cases in Adlam, without the ability to extend it later to varint

Reply:
I realized that if I want to use a high bit in the future, I can store the varint discriminant externally since there are some unused bits in #5510. So for now I'll go ahead and just use the high bit in the TinyVarVar.